pyproject.toml

```
@@ -0,0 +1,2 @@
+[tool.ruff]
+target-version = "py312"
```

What is the minimum version that CCCL aims to support? That is what should be used here.

I would encourage 3.10 (minimum for RAPIDS, [minimum for NEP29](https://numpy.org/neps/nep-0029-deprecation_policy.html)).

Other possibilities:
- Python 3.9 is [still receiving security updates until October 2025](https://endoflife.date/python)
- Python 3.11 is the [minimum for SPEC 0](https://scientific-python.org/specs/spec-0000/) but I feel that is too aggressive

```suggestion
target-version = "py310"
```

---

Done: https://github.com/NVIDIA/cccl/pull/3130/commits/8ac225484c96964e864644a587a0eb53979a7d01

It didn't change anything. TBH I'm still not sure if the pyproject.toml file actually has any effect for ruff running via pre-commit. E.g. putting the exclusions here definitely didn't work.

---

Usually the changes between Python versions are minimal-to-none.

However, I'm surprised if `pyproject.toml` settings aren't taking effect. They definitely should, from what we see in RAPIDS. You can see the [.pre-commit-config.yaml](https://github.com/rapidsai/cudf/blob/branch-25.02/.pre-commit-config.yaml) and [pyproject.toml](https://github.com/rapidsai/cudf/blob/branch-25.02/pyproject.toml#L25-L133) for an example.

Did you put the exclusions under `[tool.ruff.lint]` or `[tool.ruff]`? They should be in the `[tool.ruff.lint]` key to take effect.